File: src/main/java/gg/mineral/practice/party/Party.java
package gg.mineral.practice.party;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.Queue;
import java.util.UUID;
import java.util.concurrent.ConcurrentLinkedQueue;

import gg.mineral.practice.entity.Profile;
import gg.mineral.practice.managers.PartyManager;
import gg.mineral.practice.queue.QueueSettings;
import gg.mineral.practice.queue.QueuedEntity;
import gg.mineral.practice.util.messages.ChatMessage;
import gg.mineral.practice.util.messages.impl.ChatMessages;
import lombok.Getter;
import lombok.Setter;

@Getter
public class Party implements QueuedEntity {
	private final Profile partyLeader;
	@Setter
	private boolean open = false;
	private final ConcurrentLinkedQueue<Profile> partyMembers = new ConcurrentLinkedQueue<Profile>();

	public Party(Profile partyLeader) {
		this.partyLeader = partyLeader;
		PartyManager.registerParty(this);
	}

	public boolean equals(Party p) {
		return p.getPartyLeader().equals(partyLeader);
	}

	public void add(Profile p) {
		partyMembers.add(p);
	}

	public void remove(Profile p) {
		partyMembers.remove(p);
	}

	public boolean contains(Profile p) {
		return partyMembers.contains(p);
	}

	public void leave(Profile profile) {
		ChatMessage leftMessage = ChatMessages.LEFT_PARTY.clone().replace("%player%", profile.getName());

		Iterator<Profile> iter = getPartyMembers().iterator();

		if (getPartyLeader().equals(profile)) {
			while (iter.hasNext()) {
				Profile member = iter.next();
				iter.remove();
				member.removeFromParty();
				member.message(leftMessage);
			}

			PartyManager.remove(this);
		} else {

			profile.removeFromParty();

			while (iter.hasNext()) {
				Profile member = iter.next();
				member.message(leftMessage);
			}
		}
	}

	@Override
	public Queue<Profile> getProfiles() {
		return new LinkedList<>(partyMembers);
	}

	@Override
	public UUID getUuid() {
		return partyLeader.getUuid();
	}

	@Override
	public QueueSettings getQueueSettings() {
		return partyLeader.getQueueSettings();
	}
}

File: src/main/kotlin/gg/mineral/practice/party/Party.kt
package gg.mineral.practice.party

import gg.mineral.practice.entity.Profile
import gg.mineral.practice.managers.PartyManager.registerParty
import gg.mineral.practice.managers.PartyManager.remove
import gg.mineral.practice.queue.QueueSettings
import gg.mineral.practice.queue.QueuedEntity
import gg.mineral.practice.util.messages.impl.ChatMessages
import java.util.*


class Party(val partyLeader: Profile) : QueuedEntity {
    var open = false
    val partyMembers = mutableSetOf<Profile>()

    init {
        registerParty(this)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Party) return false
        return other.partyLeader == partyLeader
    }

    fun isPartyLeader(profile: Profile) = partyLeader == profile

    fun add(p: Profile) = partyMembers.add(p)

    fun remove(p: Profile) = partyMembers.remove(p)

    fun contains(p: Profile) = partyMembers.contains(p)

    fun leave(profile: Profile) {
        val leftMessage = ChatMessages.LEFT_PARTY.clone().replace("%player%", profile.name)

        if (partyLeader == profile) {
            partyMembers.removeIf {
                it.party = null
                it.message(leftMessage)
                true
            }
            remove(this)
        } else {
            profile.party = null
            partyMembers.forEach { it.message(leftMessage) }
        }
    }


    override fun hashCode() = javaClass.hashCode()

    override val profiles: Queue<Profile>
        get() = LinkedList(partyMembers)

    override val uuid: UUID
        get() = partyLeader.uuid

    override val queueSettings: QueueSettings
        get() = partyLeader.queueSettings
}
